UPGRADING.md

```diff
@@ -25,36 +41,56 @@ How do you know if source maps are working correctly? Try adding a syntax error
 
 ## Manifest.js
 
-Previously, if you wanted Rails to serve a non-standard named asset (any CSS not called `application.css` or JS not called `application.js`) you would have to add those files to a precompile list. For example, if you needed a marketing page with its own CSS you might add something like this:
+When compiling assets with Sprockets, Sprockets needs to decide which top-level targets to compile, usually `application.css`, `application.js`, and images.
+
+If you are using sprockets prior to 4.0, Rails will compile `application.css`, `application.js`; and *any* files found in your assets directory(ies) that are _not_ recognized as JS or CSS, but do have a filename extension. That latter was meant to apply to all your images usually in `./app/assets/images/`, but could have targetted other files as well.
+
+If you wanted to specify additional assets to deliver that were not included by this logic, for instance for a marketing page iwht its own CSS, you might add something like this:
 
 
 ```ruby
-# In your Rails configuration
+# In your Rails configuration, prior to Sprockets 4
 config.assets.precompile += ["marketing.css"]
 ```
 
-Sprockets 3 introduced the concept of a "manifest" file that could list all assets you want to make available using the `link` directive. In this case, to compile the `marketing.css` you would set precompile to:
+If you are using Sprockets 4, Rails changes it's default logic for determining top-level targets.  It will now use _only_ a file at `./app/assets/config/manifest.js` for specifying top-level targets; this file may already exist in your Rails app (although Rails only starts automatically using it once you are using sprockets 4), if not you should create it.
 
-```ruby
-config.assets.precompile = ["manifest.js"]
+The `manifest.js` file is meant to specify what files to use as a top-level target using sprockets methods `link`, `link_directory`, and `link_tree`.
+
+The default `manifest.js` created by `rails new` for the past few Rails versions looks like:
+
+```js
+//= link_tree ../images
+//= link_directory ../javascripts .js
+//= link_directory ../stylesheets .css
 ```
 
-Then you will `link` to that css file in your `manifest.js` file. In Sprockets the `link` directive declares that your file has a dependency on the thing you are linking, so it guarantees it will be compiled. Here's our example manifest file:
+This says to include the contents of all file found in the `./app/assets/images` directory or any subdirectories. And any files recognized as CSS directly at `./app/assets/stylesheets`  or as JS at `./app/assets/javascripts` (not including subdirectories). (The JS line is not generated in Rails 6.0.0 apps, since Rails 6.0 apps do not manage JS with sprockets).
```

Comment:
```suggestion
This says to include the contents of all file found in the `./app/assets/images` directory or any subdirectories. And any files recognized as CSS directly at `./app/assets/stylesheets`  or as JS at `./app/assets/javascripts` (not including subdirectories). (The JS line is not generated in Rails 6.0 apps, since Rails 6.0 apps do not manage JS with sprockets).
```